File: src/test/scala/nu/tengstrand/tetrisanalyzer/piecegenerator/PredictablePieceGeneratorTest.scala
package nu.tengstrand.tetrisanalyzer.piecegenerator

import org.junit.Test
import nu.tengstrand.tetrisanalyzer.BaseTest
import nu.tengstrand.tetrisanalyzer.piece.{PieceO, PieceI, PieceZ, PieceS}

class PredictablePieceGeneratorTest extends BaseTest {

  @Test def piece {
    val pieceGenerator = new PredictablePieceGenerator(List(new PieceS, new PieceZ))

    pieceGenerator.nextPiece() should  be (new PieceS)
    pieceGenerator.nextPiece() should  be (new PieceZ)
  }

  @Test def pieceAndNextPiece {
    val pieceGenerator = new PredictablePieceGenerator(List(new PieceS, new PieceZ, new PieceI))
    val result = pieceGenerator.nextPiece() ::
      pieceGenerator.nextPiece() ::
      pieceGenerator.nextPiece() :: Nil;

    result should be (List(new PieceS(), new PieceZ(), new PieceI()))
  }

  @Test def runOutOfPieces {
    val pieceGenerator = new PredictablePieceGenerator(List(new PieceI))
    val result = pieceGenerator.nextPiece() ::
      pieceGenerator.nextPiece() ::
      pieceGenerator.nextPiece() :: Nil;

    result should be (List(new PieceI(), new PieceO(), new PieceO()))
  }
}

File: langs/java/src/test/java/com/github/tetrisanalyzer/piecegenerator/PredictablePieceGeneratorTest.java
package com.github.tetrisanalyzer.piecegenerator;

import org.junit.Test;

import java.util.Arrays;
import java.util.List;

import static com.github.tetrisanalyzer.piece.Piece.*;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.fail;

public class PredictablePieceGeneratorTest {

    @Test
    public void nextPieceNumber() {
        PieceGenerator pieceGenerator = new PredictablePieceGenerator("SZI");
        List<Integer> result = Arrays.asList(pieceGenerator.nextPieceNumber(),
                pieceGenerator.nextPieceNumber(),
                pieceGenerator.nextPieceNumber());

        assertEquals(Arrays.asList(S, Z, I), result);
    }

    @Test
    public void runOutOfPieces() {
        PieceGenerator pieceGenerator = new PredictablePieceGenerator("I");
        List<Integer> result = Arrays.asList(
                pieceGenerator.nextPieceNumber(),
                pieceGenerator.nextPieceNumber(),
                pieceGenerator.nextPieceNumber(),
                pieceGenerator.nextPieceNumber(),
                pieceGenerator.nextPieceNumber());

        assertEquals(Arrays.asList(I, O, O, O, O), result);
    }

    @Test
    public void illegalPiece() {
        PieceGenerator pieceGenerator = new PredictablePieceGenerator("Sq");
        try {
            Arrays.asList(pieceGenerator.nextPieceNumber(),
                    pieceGenerator.nextPieceNumber(),
                    pieceGenerator.nextPieceNumber());
            fail();
        } catch (IllegalArgumentException e) {
            assertEquals("Illegal board character 'q', expected: -OISZLJTx+", e.getMessage());
        }
    }
}
